std/haxe/coro/schedulers/EventLoopScheduler.hx

```diff
@@ -1,24 +1,110 @@
 package haxe.coro.schedulers;
 
-import haxe.coro.EventLoop;
+import haxe.exceptions.ArgumentException;
+
+private class ScheduledEvent {
+	public final func : ()->Void;
+	public final runTime : Float;
+	public var next : Null<ScheduledEvent>;
+	public var previous : Null<ScheduledEvent>;
+
+	public function new(func, runTime) {
+		this.func    = func;
+		this.runTime = runTime;
+
+		next     = null;
+		previous = null;
+	}
+}
 
 class EventLoopScheduler extends Scheduler {
-	final loop:EventLoop;
+	var first : Null<ScheduledEvent>;
+	var last : Null<ScheduledEvent>;
+	var zeroEvents : Array<() -> Void>;
 
-	public function new(loop:EventLoop) {
+	public function new() {
 		super();
-		this.loop = loop;
+
+		first = null;
+		last = null;
+		zeroEvents = [];
 	}
 
-	public function schedule(func:() -> Void) {
-		loop.run(func);
+    public function schedule(func:()->Void, ms:Int) {
+		if (ms < 0) {
+			throw new ArgumentException("Time must be greater or equal to zero");
+		} else if (ms == 0) {
+			zeroEvents.push(func);
+			return;
+		}
+
+		final event = new ScheduledEvent(func, now() + (ms / 1000));
+		if (first == null) {
+			first = event;
+			last = event;
+			return;
+		}
+
+		var current = last;
+		while (true) {
+			if (current == null) {
+				event.next = first;
+				first = event;
+				break;
+			} else if (event.runTime >= current.runTime) {
+				final next = current.next;
+				current.next = event;
+				event.previous = current;
+				if (next != null) {
+					event.next = next;
+					next.previous = event;
+				} else {
+					last = event;
+				}
+				break;
+			} else {
+				current = current.previous;
+			}
+		}
+    }
+
+	public function now() {
+		return Timer.stamp();
 	}
 
-	public function scheduleIn(func:() -> Void, ms:Int) {
-		loop.runIn(func, ms);
+	public function run() {
+
+		while (true) {
+			final events = zeroEvents;
+			zeroEvents = [];
+			for (event in events) {
+				event();
+			}
```

Comment:
I'm never sure if it's best to do this sort of re-assign with an empty array thing or not (guess it depends what we're optimising for). In hxcpp atleast, shifting elements until empty will not shrink the underlying storage, so pushing back to the array doesn't cause it to expand as it would with a fresh array.

Reply:
Ah, right, modify while iterate.
I had an attempt at double buffering that zero events array. So when you're iterating over one it's writing to the other, and calling `flip` returns the array you can now consume from. When swapping the buffer, the one which will be written into from now on is zeroed so targets which don't shrink the underlying storage might see much lower allocations from this.
Not sure if this is any better!

Reply:
Ah that's clever, nice!